File: submit/prj1-sol/py/tl_lexer.py
#!/usr/bin/env python3

import re
import sys
from collections import namedtuple
import json

def scan_text(text):
    def next_match(text):
        m = re.compile(r'\s+').match(text)
        if (m): return (m, None)
        m = re.compile(r'#.*').match(text)
        if (m): return (m, None)
        m = re.compile(r'\d+').match(text)
        if (m): return (m, 'INT')
        m = re.compile(r'<=|>=|==|!=').match(text)
        if (m): return (m, m.group())
        m = re.compile(r'[a-zA-Z_]\w*').match(text)
        if (m): return (m, 'DEF' if m.group() == 'def' else 'ID')
        m = re.compile(r'.').match(text)  #must be last: match any char
        if (m): return (m, m.group())

    tokens = []
    while (len(text) > 0):
        (match, kind) = next_match(text)
        lexeme = match.group()
        if (kind): tokens.append(Token(kind, lexeme))
        text = text[len(lexeme):]
    tokens.append(Token('<EOF>', '<EOF>'));
    return tokens

def scan_stdin():
    return scan_text(sys.stdin.read())

def main():
    tokens = scan_stdin()
    #convert to dict to facilitate json in desired format
    jsonTokens = [{ "kind": t.kind, "lexeme": t.lexeme } for t in tokens];
    print(json.dumps(jsonTokens, separators=(',', ':'))) #no whitespace

Token = namedtuple('Token', ['kind', 'lexeme'])

if __name__ == "__main__":
    main()

File: submit/prj1-sol/js/tl-lexer.mjs
#!/usr/bin/env node

import fs from 'fs';
import Path from 'path';

/** Returns single char tokens with kind set to the char; i.e.
 *  '+' converts to Token('+', '+') with kind '+' and lexeme '+'.
 */
function scan(str) {
  const toks = [];
  let m;
  for (let s = str; s.length > 0; s = s.slice(m[0].length)) {
    if (m = s.match(/^\s+/)) { //s starts with whitespace
      continue; //skip whitespace
    }
    else if (m = s.match(/^#.*/)) {
      continue; //skip comments
    }
    else if (m = s.match(/(^<=)|(^>=)|(^==)|(^!=)/)) {
      toks.push(new Token(m[0], m[0]));
    }
    else if (m = s.match(/^\d+/)) { //one or more digits
      toks.push(new Token('INT', m[0]));
    }
    else if (m = s.match(/^[_a-zA-Z]\w*/)) {
      toks.push(new Token(m[0] === 'def' ? 'DEF' : 'ID', m[0]));
    }
    else if (m = s.match(/^./)) {  //any single char
      toks.push(new Token(m[0], m[0]));
    }
  }
  toks.push(new Token('<EOF>', '<EOF>'));
  return toks;
}

class Token {
  constructor(kind, lexeme) {
    Object.assign(this, {kind, lexeme});
  }
}

const CHAR_SET = 'utf8'
function main() {
  const text = fs.readFileSync(0, CHAR_SET);
  console.log(JSON.stringify(scan(text)));
}

function isRun() {
  const url = new URL(import.meta.url);
  const path = url.pathname;
  return (Path.basename(path) === Path.basename(process.argv[1]));
}
if (isRun()) main();

export default scan;

